### PR TITLE
Update to Pasta to make it possible to run it from python 3 (with typ…

### PR DESCRIPTION
**Pull request by bronevet-abc**
…e checking, etc.) and have it parse either python 2 or 3.

Notes:
- I've introduced variable py_ver, set to either 'PY27' or 'PY3'. We probably want to replace it with a dict {major: #, minor: #} or something like that.
- I've had to wrap a bunch of code within function calls that introduce py_ver into existing files. It causes artificially many change lines according to Critique and may look a bit weird but it seems to be the simplest way to introduce this detail within the context of the testing framework.
- I had to mutilate the check for Python > 3.5 to make it work for now but we'll correct it once we decide how to describe Python versions.
- Fixed a few bugs in the code:
  - Case like 'globals a,b,c' where the first identifier appears again in the list.
  - Support for ur prefix for Python 2 string literals

### CODE REVIEW

File: pasta/base/annotate.py [hunk 1]
```diff
@@ -28,28 +28,43 @@
 from six.moves import zip
 import sys
 import token
+import typed_ast
```
Comment (soupytwist) [inline pasta/base/annotate.py:31]: Remove `ast` import?

File: pasta/base/annotate.py [hunk 2]
```diff
@@ -28,28 +28,43 @@
...
-                 semicolon=False, comment=False, statement=False):
+
+def _gen_wrapper(f,
+                 scope=True,
+                 prefix=True,
+                 suffix=True,
+                 max_suffix_lines=None,
+                 semicolon=False,
+                 comment=False,
+                 statement=False):
```
Comment (soupytwist) [inline pasta/base/annotate.py:52]:
```suggestion
def _gen_wrapper(f, scope=True, prefix=True, suffix=True, max_suffix_lines=None,
                 semicolon=False, comment=False, statement=False):
```

File: pasta/base/annotate.py [hunk 3]
```diff
@@ -28,28 +28,43 @@
...
       f(self, node, *args, **kwargs)
       if suffix:
-        self.suffix(node, max_lines=max_suffix_lines, semicolon=semicolon,
-                    comment=comment, default='\n' if statement else '')
+        self.suffix(
+            node,
+            max_lines=max_suffix_lines,
+            semicolon=semicolon,
+            comment=comment,
+            default='\n' if statement else '')
```
Comment (soupytwist) [inline pasta/base/annotate.py:66]:
```suggestion
        self.suffix(node, max_lines=max_suffix_lines, semicolon=semicolon,
                    comment=comment, default='\n' if statement else '')
```

File: pasta/base/annotate.py [hunk 4]
```diff
@@ -80,8 +95,13 @@ def space_left(f):
...
   """Decorates a function where the node is a statement."""
-  return _gen_wrapper(f, scope=False, max_suffix_lines=1, semicolon=True,
-                      comment=True, statement=True)
+  return _gen_wrapper(
+      f,
+      scope=False,
+      max_suffix_lines=1,
+      semicolon=True,
+      comment=True,
+      statement=True)
```
Comment (soupytwist) [inline pasta/base/annotate.py:104]:
```suggestion
  return _gen_wrapper(f, scope=False, max_suffix_lines=1, semicolon=True,
                      comment=True, statement=True)
```

File: pasta/base/annotate.py [hunk 5]
```diff
@@ -103,1433 +124,1594 @@ def wrapped(self, node, *args, **kwargs):
...
-
-    Arguments:
-      max_lines: (int) Maximum number of newlines to consider.
-      semicolon: (boolean) If True, parse up to the next semicolon (if present).
-      comment: (boolean) If True, look for a trailing comment even when not in
-        a parenthesized scope.
-    """
-    return ''
+  class BaseVisitor(ast27.NodeVisitor if py_ver < (3, 0) else ast3.NodeVisitor):
+    # class BaseVisitor(ast.NodeVisitor):
```
Comment (soupytwist) [inline pasta/base/annotate.py:139]:
```suggestion
```

File: pasta/base/annotate.py [hunk 6]
```diff
@@ -103,1433 +124,1594 @@ def wrapped(self, node, *args, **kwargs):
...
+    """Walks a syntax tree in the order it appears in code.
 
-  def dots(self, num_dots):
-    """Account for a number of dots."""
-    return '.' * num_dots
+    This class has a dual-purpose. It is implemented (in this file) for
```
Comment (soupytwist) [inline pasta/base/annotate.py:142]: Please fix text wrapping

Comment (soupytwist) [reply pasta/base/annotate.py:142]: (here and elsewhere in this file)

File: pasta/base/annotate.py [hunk 7]
```diff
@@ -103,1433 +124,1594 @@ def wrapped(self, node, *args, **kwargs):
...
-  def check_is_elif(self, node):
-    """Return True if the node continues a previous `if` statement as `elif`.
+          node, 'prefix', [lambda: self.ws(comment=True)], default=default)
+
+    def suffix(self,
+               node,
+               max_lines=None,
+               semicolon=False,
+               comment=False,
+               default=''):
```
Comment (soupytwist) [inline pasta/base/annotate.py:177]:
```suggestion
      def suffix(self, node, max_lines=None, semicolon=False, comment=False,
             default=''):
```

File: pasta/base/annotate.py [hunk 8]
```diff
@@ -103,1433 +124,1594 @@ def wrapped(self, node, *args, **kwargs):
...
+
+    def set_default_indent_diff(self, indent):
+      self._default_indent_diff = indent
+
+    @contextlib.contextmanager
+    def scope(self,
+              node,
+              attr=None,
+              trailing_comma=False,
+              default_parens=False):
```
Comment (soupytwist) [inline pasta/base/annotate.py:209]:
```suggestion
    def scope(self, node, attr=None, trailing_comma=False, default_parens=False):
```

File: pasta/base/annotate.py [hunk 9]
```diff
@@ -103,1433 +124,1594 @@ def wrapped(self, node, *args, **kwargs):
...
+      self.attr(node, 'names', identifiers)
+
+    @statement
+    def visit_Import(self, node):
+      self.token('import')
       for i, alias in enumerate(node.names):
         self.attr(node, 'alias_prefix_%d' % i, [self.ws], default=' ')
         self.visit(alias)
-        if alias is not node.names[-1]:
+        if alias != node.names[-1]:
```
Comment (soupytwist) [inline pasta/base/annotate.py:747]: Do you remember why this is needed? Alias is an ast node, isn't it?

File: pasta/base/annotate.py [hunk 10]
```diff
@@ -103,1433 +124,1594 @@ def wrapped(self, node, *args, **kwargs):
...
+        for part in parts[:-1]:
+          module_pattern += [self.ws, part, self.ws, '.']
+        module_pattern += [self.ws, parts[-1]]
+
+      self.attr(
+          node,
+          'module',
+          module_pattern,
+          deps=('level', 'module'),
+          default='.' * node.level + (node.module or ''))
```
Comment (soupytwist) [inline pasta/base/annotate.py:769]:
```suggestion
      self.attr(node, 'module', module_pattern, deps=('level', 'module'),
                default='.' * node.level + (node.module or ''))
```

File: pasta/base/annotate.py [hunk 11]
```diff
@@ -103,1433 +124,1594 @@ def wrapped(self, node, *args, **kwargs):
...
+    @expression
+    def visit_IfExp(self, node):
+      self.visit(node.body)
+      self.attr(node, 'if', [self.ws, 'if', self.ws], default=' if ')
+      self.visit(node.test)
+      self.attr(
+          node,
+          'else', [self.ws, 'else', self.ws],
+          default=' else ',
+          separate_before=True)
```
Comment (soupytwist) [inline pasta/base/annotate.py:1046]:
```suggestion
      self.attr(node, 'else', [self.ws, 'else', self.ws], default=' else ',
                separate_before=True)
```

File: pasta/base/annotate.py [hunk 12]
```diff
@@ -103,1433 +124,1594 @@ def wrapped(self, node, *args, **kwargs):
...
+        self.tokens.hint_closed()
+
+      return token.src
+
+    def optional_token(self,
+                       node,
+                       attr_name,
+                       token_val,
+                       allow_whitespace_prefix=False,
+                       default=False):
```
Comment (soupytwist) [inline pasta/base/annotate.py:1619]:
```suggestion
    def optional_token(self, node, attr_name, token_val,
                       allow_whitespace_prefix=False, default=False):
```

File: pasta/base/annotate_test.py [hunk 13]
```diff
@@ -28,6 +28,8 @@
 from six import with_metaclass
 import sys
 import textwrap
+from typed_ast import ast27
```
Comment (soupytwist) [inline pasta/base/annotate_test.py:31]: `_ast` and `ast` imports -- still needed?

**Review by soupytwist — CHANGES_REQUESTED** [review]
I left a lot of suggestions -- not sure how easy it is to incorporate these with GH's UI, hopefully it helps. Most of it is formatting, but please read. Overall looks good! I do want to make sure that it can still be used without ever specifying a python version (using the default `sys.version_info[:2]`).

File: pasta/augment/import_utils.py [hunk 14]
```diff
@@ -61,41 +69,52 @@ def add_import(tree, name_to_import, asname=None, from_import=True, merge_from_i
 
   import_node = None
   added_name = None
-  
+
   def make_safe_alias_node(alias_name, asname):
     # Try to avoid name conflicts
-    new_alias = ast.alias(name=alias_name, asname=asname)
+    if py_ver < (3, 0):
```
[the remark this answers is not in the document]
Comment (bronevet-abc) [reply pasta/augment/import_utils.py:71]: Done. Added a method to return ast27 or ast3. Using it in all similar code.

File: pasta/augment/import_utils.py [hunk 15]
```diff
@@ -147,10 +166,10 @@ def get_unused_import_aliases(tree, sc=None):
     the given tree.
   """
   if sc is None:
-    sc = scope.analyze(tree)
+    sc = scope.analyze(tree, py_ver)
   unused_aliases = set()
-  for node in ast.walk(tree):
-    if isinstance(node, ast.alias):
+  for node in pasta.ast_walk(tree, py_ver):
+    if isinstance(node, (ast27.alias, ast3.alias)):
```
[the remark this answers is not in the document]
Comment (bronevet-abc) [reply pasta/augment/import_utils.py:168]: Couldn't think of anything clean

File: pasta/augment/import_utils_test.py [hunk 16]
```diff
@@ -29,393 +31,421 @@
...
 import b
 a.foo()
-"""
-    tree = ast.parse(src)
-    self.assertItemsEqual(import_utils.get_unused_import_aliases(tree),
-                          [tree.body[1].names[0]])
-
-  def test_import_from(self):
-    src = """\
+  """
```
[the remark this answers is not in the document]
Comment (bronevet-abc) [reply pasta/augment/import_utils_test.py:139]: Done

File: pasta/augment/rename.py [hunk 17]
```diff
@@ -20,13 +20,16 @@
...
+from typed_ast import ast3
 
+import pasta
 from pasta.augment import import_utils
 from pasta.base import ast_utils
 from pasta.base import scope
 
 
-def rename_external(t, old_name, new_name):
+def rename_external(t, old_name, new_name, py_ver):
```
[the remark this answers is not in the document]
Comment (bronevet-abc) [reply pasta/augment/rename.py:33]: Done

File: pasta/augment/rename.py [hunk 18]
```diff
@@ -56,7 +59,7 @@ def rename_external(t, old_name, new_name):
   Returns:
     True if any changes were made, False otherwise.
   """
-  sc = scope.analyze(t)
+  sc = scope.analyze(t, py_ver)
```
[the remark this answers is not in the document]
Comment (bronevet-abc) [reply pasta/augment/rename.py:63]: Done

File: pasta/base/ast_utils.py [hunk 19]
```diff
@@ -19,6 +19,9 @@
 from __future__ import print_function
 
 import ast
```
[the remark this answers is not in the document]
Comment (bronevet-abc) [reply pasta/base/ast_utils.py:21]: Done

**Review by soupytwist — APPROVED** [review]

**Pull request merged**